tests/php/test-class-amp-youtube-embed-handler.php

```diff
@@ -66,4 +124,96 @@ public function test_video_override() {
 		remove_all_filters( 'wp_video_shortcode_override' );
 	}
 
+	public function get_conversion_data() {
+		return [
+			'no_embed'                         => [
+				'<p>Hello world.</p>',
+				'<p>Hello world.</p>' . PHP_EOL,
+			],
+
+			'url_simple'                       => [
+				'https://www.youtube.com/watch?v=kfVsfOSbJY0' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270" title="Rebecca Black - Friday"><a fallback href="https://www.youtube.com/watch?v=kfVsfOSbJY0">Rebecca Black &#8211; Friday</a></amp-youtube></p>' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270"></amp-youtube></p>' . PHP_EOL,
+			],
+
+			'url_short'                        => [
+				'https://youtu.be/kfVsfOSbJY0' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270" title="Rebecca Black - Friday"><a fallback href="https://youtu.be/kfVsfOSbJY0">Rebecca Black &#8211; Friday</a></amp-youtube></p>' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270"></amp-youtube></p>' . PHP_EOL,
+			],
+
+			'url_with_querystring'             => [
+				'http://www.youtube.com/watch?v=kfVsfOSbJY0&hl=en&fs=1&w=425&h=349' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270" title="Rebecca Black - Friday"><a fallback href="http://www.youtube.com/watch?v=kfVsfOSbJY0&#038;hl=en&#038;fs=1&#038;w=425&#038;h=349">Rebecca Black &#8211; Friday</a></amp-youtube></p>' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270"></amp-youtube></p>' . PHP_EOL,
+			],
+
+			// Several reports of invalid URLs that have multiple `?` in the URL.
+			'url_with_querystring_and_extra_?' => [
+				'http://www.youtube.com/watch?v=kfVsfOSbJY0?hl=en&fs=1&w=425&h=349' . PHP_EOL,
+				'<p>http://www.youtube.com/watch?v=kfVsfOSbJY0?hl=en&#038;fs=1&#038;w=425&#038;h=349</p>' . PHP_EOL,
+			],
+
+			'shortcode_unnamed_attr_as_url'    => [
+				'[youtube http://www.youtube.com/watch?v=kfVsfOSbJY0]' . PHP_EOL,
+				'<amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="600" height="338"></amp-youtube>' . PHP_EOL,
+			],
+
+			'embed_url'                        => [
+				'https://www.youtube.com/embed/kfVsfOSbJY0' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270" title="Rebecca Black - Friday"><a fallback href="https://youtube.com/watch?v=kfVsfOSbJY0">Rebecca Black &#8211; Friday</a></amp-youtube></p>' . PHP_EOL,
+				'<p><amp-youtube data-videoid="kfVsfOSbJY0" layout="responsive" width="480" height="270"></amp-youtube></p>' . PHP_EOL,
+			],
+		];
+	}
+
+	/**
+	 * @dataProvider get_conversion_data
+	 */
+	public function test__conversion( $source, $expected, $fallback = null ) {
```

Comment:
Could you rename `$fallback` to be something else here? It gets confusing with `fallback` content in AMP. Perhaps `$expected_without_fallback` would be better.

Reply:
Ah, OK. `$expected_without_fallback` seems a bit confusing as well, how about something direct like `$fallback_for_expected`?